File: hdf-converters/converters/semgrep-to-hdf/typescript/converter.test.ts
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { describe, it, expect } from 'vitest';
import { convertSemgrepToHdf } from './converter.js';
import { runConverterContractTests } from '../../../shared/typescript/converter-contract.js';
import { expectValidResults } from '../../../test/helpers/expectValidHdf.js';
import { ResultStatus } from '@mitre/hdf-schema';
import type { HDFResults, EvaluatedRequirement } from '@mitre/hdf-schema';

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURES_DIR = join(__dirname, '..', 'fixtures');

function loadFixture(name: string): string {
  return readFileSync(join(FIXTURES_DIR, 'input', name), 'utf-8');
}

async function convert(name: string): Promise<HDFResults> {
  return JSON.parse(await convertSemgrepToHdf(loadFixture(name))) as HDFResults;
}

function findReq(hdf: HDFResults, idFragment: string): EvaluatedRequirement | undefined {
  return hdf.baselines[0]!.requirements.find((r) => r.id.includes(idFragment));
}

runConverterContractTests({
  converterName: 'semgrep-to-hdf',
  convertFn: convertSemgrepToHdf,
  minimalFixture: 'minimal.json',
});

describe('semgrep to HDF converter', () => {
  it('rejects input that is not a Semgrep report', async () => {
    await expect(convertSemgrepToHdf('{"foo":1}')).rejects.toThrow(
      'does not look like a Semgrep report',
    );
  });

  describe('real fixture', () => {
    it('produces a schema-valid, correctly-shaped document', async () => {
      const hdf = await convert('real.json');
      expectValidResults(hdf);

      const bl = hdf.baselines[0]!;
      expect(bl.name).toBe('Semgrep Scan');
      expect(hdf.tool?.name).toBe('Semgrep');
      expect(hdf.tool?.version).toBe('1.174.0');
      expect(hdf.generator?.name).toBe('semgrep-to-hdf');
    });

    it('groups findings into one requirement per rule', async () => {
      const hdf = await convert('real.json');
      const ids = hdf.baselines[0]!.requirements.map((r) => r.id);
      expect(new Set(ids).size).toBe(ids.length);
      expect(ids).toHaveLength(2);
    });

    it('maps semgrep severity onto impact', async () => {
      const hdf = await convert('real.json');
      // ERROR severity
      expect(findReq(hdf, 'subprocess-shell-true')?.impact).toBe(0.7);
      // WARNING severity
      expect(findReq(hdf, 'dynamic-urllib-use-detected')?.impact).toBe(0.5);
    });

    it('resolves NIST tags from the rule CWE', async () => {
      const hdf = await convert('real.json');
      const req = findReq(hdf, 'subprocess-shell-true');
      expect(req?.tags?.nist).toBeDefined();
      expect(Array.isArray(req?.tags?.nist)).toBe(true);
      expect((req?.tags?.nist as string[]).length).toBeGreaterThan(0);
    });

    it('derives CCI tags from the resolved NIST controls', async () => {
      const hdf = await convert('real.json');
      const req = findReq(hdf, 'subprocess-shell-true')!;
      expect(Array.isArray(req.tags?.cci)).toBe(true);
      expect((req.tags?.cci as string[]).length).toBeGreaterThan(0);
      for (const cci of req.tags?.cci as string[]) {
        expect(cci).toMatch(/^CCI-\d+$/);
      }
    });

    it('normalizes owasp whether the rule supplies a string or an array', async () => {
      const hdf = await convert('real.json');
      for (const req of hdf.baselines[0]!.requirements) {
        expect(Array.isArray(req.tags?.owasp)).toBe(true);
      }
      // real.json deliberately carries one of each form
      expect((findReq(hdf, 'subprocess-shell-true')?.tags?.owasp as string[]).length).toBe(3);
      expect((findReq(hdf, 'dynamic-urllib-use-detected')?.tags?.owasp as string[]).length).toBe(1);
    });

    it('does not let semgrep metadata impact shadow the HDF impact float', async () => {
      const hdf = await convert('real.json');
      const req = findReq(hdf, 'subprocess-shell-true')!;
      expect(typeof req.impact).toBe('number');
      expect(req.tags?.impact).toBeUndefined();
      expect(req.tags?.semgrepImpact).toBe('LOW');
    });

    it('preserves the cross-framework metadata SARIF drops', async () => {
      const hdf = await convert('real.json');
      const req = findReq(hdf, 'dynamic-urllib-use-detected')!;
      expect(req.tags?.likelihood).toBe('LOW');
      expect(req.tags?.confidence).toBe('LOW');
      expect(req.tags?.asvs).toBeDefined();
      expect(req.tags?.vulnerabilityClass).toBeDefined();
    });

    it('reports every finding as failed, since semgrep omits suppressed findings', async () => {
      const hdf = await convert('real.json');
      for (const req of hdf.baselines[0]!.requirements) {
        for (const result of req.results) {
          expect(result.status).toBe(ResultStatus.Failed);
        }
      }
    });

    it('never emits the redacted placeholder semgrep uses in unauthenticated scans', async () => {
      const hdf = await convert('real.json');
      expect(JSON.stringify(hdf)).not.toContain('requires login');
    });

    it('records the finding location in codeDesc', async () => {
      const hdf = await convert('real.json');
      const result = findReq(hdf, 'subprocess-shell-true')!.results[0]!;
      expect(result.codeDesc).toContain('app/handlers.py');
      expect(result.codeDesc).toContain('7');
    });
  });

  describe('empty fixture', () => {
    it('produces a no-findings requirement rather than an empty baseline', async () => {
      const hdf = await convert('empty.json');
      expectValidResults(hdf);
      expect(hdf.baselines[0]!.requirements).toHaveLength(1);
      expect(hdf.baselines[0]!.requirements[0]!.impact).toBe(0);
    });
  });

  describe('errors fixture', () => {
    it('surfaces scan errors as their own requirement with error status', async () => {
      const hdf = await convert('errors.json');
      expectValidResults(hdf);
      const req = findReq(hdf, 'semgrep-scan-errors');
      expect(req).toBeDefined();
      expect(req!.results.length).toBe(2);
      for (const result of req!.results) {
        expect(result.status).toBe(ResultStatus.Error);
      }
    });

    it('omits the scan-errors requirement when the scan reported none', async () => {
      const hdf = await convert('real.json');
      expect(findReq(hdf, 'semgrep-scan-errors')).toBeUndefined();
    });
  });
});

describe('sparse and malformed rules', () => {
  function scan(results: unknown[], extra: Record<string, unknown> = {}): string {
    return JSON.stringify({
      version: '1.174.0',
      results,
      errors: [],
      paths: { scanned: ['a.py'] },
      engine_requested: 'OSS',
      ...extra,
    });
  }

  it('converts a rule that declares no metadata at all', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(scan([{ check_id: 'bare.rule', path: 'a.py' }])),
    ) as HDFResults;
    const req = hdf.baselines[0]!.requirements[0]!;
    // Falls back to the static-analysis default tags.
    expect(req.tags?.nist).toEqual(['SA-11', 'RA-5']);
    expect(req.tags?.owasp).toBeUndefined();
    expect(req.tags?.confidence).toBeUndefined();
    expect(req.tags?.likelihood).toBeUndefined();
    expect(req.tags?.semgrepImpact).toBeUndefined();
    expect(req.tags?.category).toBeUndefined();
    expect(req.tags?.banditCode).toBeUndefined();
    expect(req.tags?.asvs).toBeUndefined();
    expect(req.tags?.references).toBeUndefined();
    expect(req.tags?.subcategory).toBeUndefined();
    expect(req.tags?.technology).toBeUndefined();
    expect(req.tags?.vulnerabilityClass).toBeUndefined();
    // No location fields at all.
    expect(req.results[0]!.codeDesc).toBe('Path: a.py');
    expect(req.results[0]!.message).toBe('');
  });

  it('falls back to moderate impact for an unknown or absent severity', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        scan([
          { check_id: 'a', extra: { severity: 'NOVEL' } },
          { check_id: 'b' },
          { check_id: 'c', extra: { severity: 'CRITICAL' } },
        ]),
      ),
    ) as HDFResults;
    const reqs = hdf.baselines[0]!.requirements;
    expect(reqs.find((r) => r.id === 'a')!.impact).toBe(0.5);
    expect(reqs.find((r) => r.id === 'b')!.impact).toBe(0.5);
    expect(reqs.find((r) => r.id === 'c')!.impact).toBe(0.9);
  });

  it('renders a multi-line span and a path-only location', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        scan([
          { check_id: 'span', path: 'a.py', start: { line: 4 }, end: { line: 9 } },
          { check_id: 'single', path: 'a.py', start: { line: 4 }, end: { line: 4 } },
          { check_id: 'nopath' },
        ]),
      ),
    ) as HDFResults;
    const reqs = hdf.baselines[0]!.requirements;
    expect(reqs.find((r) => r.id === 'span')!.results[0]!.codeDesc).toBe('Path: a.py, lines 4-9');
    expect(reqs.find((r) => r.id === 'single')!.results[0]!.codeDesc).toBe('Path: a.py, line 4');
    expect(reqs.find((r) => r.id === 'nopath')!.results[0]!.codeDesc).toBe('Path: unknown');
  });

  it('drops redacted fields but keeps real ones', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        scan([
          { check_id: 'redacted', extra: { lines: 'requires login', fingerprint: 'requires login' } },
          { check_id: 'real', extra: { lines: 'x = 1' } },
        ]),
      ),
    ) as HDFResults;
    const reqs = hdf.baselines[0]!.requirements;
    expect(reqs.find((r) => r.id === 'redacted')!.results[0]!.message).toBe('');
    expect(reqs.find((r) => r.id === 'real')!.results[0]!.message).toContain('Matched code:');
  });

  it('skips findings with no check_id and falls back to the no-findings placeholder', async () => {
    const hdf = JSON.parse(await convertSemgrepToHdf(scan([{ path: 'a.py' }]))) as HDFResults;
    expect(hdf.baselines[0]!.requirements).toHaveLength(1);
    expect(hdf.baselines[0]!.requirements[0]!.id).toBe('semgrep-no-findings');
  });

  it('collapses repeated occurrences of one rule into a single requirement', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        scan([
          { check_id: 'dup', path: 'a.py', start: { line: 1 }, end: { line: 1 } },
          { check_id: 'dup', path: 'b.py', start: { line: 2 }, end: { line: 2 } },
        ]),
      ),
    ) as HDFResults;
    expect(hdf.baselines[0]!.requirements).toHaveLength(1);
    expect(hdf.baselines[0]!.requirements[0]!.results).toHaveLength(2);
  });

  it('deduplicates reference urls drawn from several metadata fields', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        scan([
          {
            check_id: 'refs',
            extra: {
              metadata: {
                references: ['https://a', 'https://a'],
                source: 'https://a',
                shortlink: 'https://b',
                'source-rule-url': '',
                asvs: { control_url: 'https://c' },
              },
            },
          },
        ]),
      ),
    ) as HDFResults;
    expect(hdf.baselines[0]!.requirements[0]!.tags?.references).toEqual([
      'https://a',
      'https://b',
      'https://c',
    ]);
  });

  it('ignores a CWE entry that carries no parsable id', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        scan([{ check_id: 'badcwe', extra: { metadata: { cwe: ['not a cwe reference'] } } }]),
      ),
    ) as HDFResults;
    expect(hdf.baselines[0]!.requirements[0]!.tags?.nist).toEqual(['SA-11', 'RA-5']);
  });

  it('handles scan errors whose type is a bare string or absent', async () => {
    const hdf = JSON.parse(
      await convertSemgrepToHdf(
        JSON.stringify({
          version: '1.174.0',
          results: [],
          errors: [
            { message: 'a', type: 'PlainString', path: 'x.yml' },
            { message: 'b' },
          ],
          paths: { scanned: [] },
        }),
      ),
    ) as HDFResults;
    const req = hdf.baselines[0]!.requirements.find((r) => r.id === 'semgrep-scan-errors')!;
    expect(req.results[0]!.message).toContain('PlainString');
    expect(req.results[1]!.message).toContain('Unknown');
    expect(req.results[1]!.codeDesc).toBe('Path: unknown');
  });

  it('derives a title from a single-segment rule id', async () => {
    const hdf = JSON.parse(await convertSemgrepToHdf(scan([{ check_id: 'rule' }]))) as HDFResults;
    expect(hdf.baselines[0]!.requirements[0]!.title).toBe('Rule');
  });
});

File: hdf-converters/converters/semgrep-to-hdf/go/converter_test.go
package semgrep

import (
	"os"
	"path/filepath"
	"testing"
	"time"

	shared "github.com/mitre/hdf-libs/hdf-converters/v3/shared/go"
	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

const testVersion = "1.0.0"

func loadFixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "fixtures", "input", name))
	require.NoError(t, err, "failed to read fixture %s", name)
	return data
}

func convertFixture(t *testing.T, name string) *hdf.HDFResults {
	t.Helper()
	out, err := ConvertSemgrepToHDF(loadFixture(t, name), testVersion)
	require.NoError(t, err)
	require.NotNil(t, out)
	return out
}

func findReq(reqs []hdf.EvaluatedRequirement, fragment string) *hdf.EvaluatedRequirement {
	for i := range reqs {
		if len(fragment) > 0 && contains(reqs[i].ID, fragment) {
			return &reqs[i]
		}
	}
	return nil
}

func contains(haystack, needle string) bool {
	return len(needle) <= len(haystack) && (func() bool {
		for i := 0; i+len(needle) <= len(haystack); i++ {
			if haystack[i:i+len(needle)] == needle {
				return true
			}
		}
		return false
	})()
}

// ---- Contract ----

func TestRejectsEmptyInput(t *testing.T) {
	_, err := ConvertSemgrepToHDF([]byte(""), testVersion)
	require.Error(t, err)
}

func TestRejectsInvalidJSON(t *testing.T) {
	_, err := ConvertSemgrepToHDF([]byte("not valid json"), testVersion)
	require.Error(t, err)
}

func TestRejectsNonSemgrepDocument(t *testing.T) {
	_, err := ConvertSemgrepToHDF([]byte(`{"foo":1}`), testVersion)
	require.ErrorContains(t, err, "does not look like a Semgrep report")
}

func TestConvertsMinimalFixture(t *testing.T) {
	out := convertFixture(t, "minimal.json")
	require.Len(t, out.Baselines, 1)
	assert.Equal(t, "Semgrep Scan", out.Baselines[0].Name)
}

// ---- Mapping behaviour ----

func TestGroupsFindingsOneRequirementPerRule(t *testing.T) {
	out := convertFixture(t, "real.json")
	reqs := out.Baselines[0].Requirements
	assert.Len(t, reqs, 2)
	seen := map[string]bool{}
	for _, r := range reqs {
		assert.False(t, seen[r.ID], "duplicate requirement id %s", r.ID)
		seen[r.ID] = true
	}
}

func TestMapsSeverityToImpact(t *testing.T) {
	out := convertFixture(t, "real.json")
	reqs := out.Baselines[0].Requirements
	assert.Equal(t, 0.7, findReq(reqs, "subprocess-shell-true").Impact)
	assert.Equal(t, 0.5, findReq(reqs, "dynamic-urllib-use-detected").Impact)
}

func TestResolvesNistAndCciTags(t *testing.T) {
	out := convertFixture(t, "real.json")
	req := findReq(out.Baselines[0].Requirements, "subprocess-shell-true")
	require.NotNil(t, req)
	nist := shared.NISTTagsFromMap(req.Tags)
	assert.NotEmpty(t, nist)
	ccis, ok := req.Tags["cci"].([]string)
	require.True(t, ok, "cci tag missing or wrong type")
	assert.NotEmpty(t, ccis)
}

func TestNormalizesOwaspStringOrArray(t *testing.T) {
	out := convertFixture(t, "real.json")
	reqs := out.Baselines[0].Requirements
	// real.json deliberately carries one rule with an array and one with a string
	arrayForm, ok := findReq(reqs, "subprocess-shell-true").Tags["owasp"].([]string)
	require.True(t, ok)
	assert.Len(t, arrayForm, 3)
	stringForm, ok := findReq(reqs, "dynamic-urllib-use-detected").Tags["owasp"].([]string)
	require.True(t, ok)
	assert.Len(t, stringForm, 1)
}

func TestSemgrepMetadataImpactDoesNotShadowHdfImpact(t *testing.T) {
	out := convertFixture(t, "real.json")
	req := findReq(out.Baselines[0].Requirements, "subprocess-shell-true")
	require.NotNil(t, req)
	_, shadowed := req.Tags["impact"]
	assert.False(t, shadowed, "semgrep metadata.impact must not be tagged as impact")
	assert.Equal(t, "LOW", req.Tags["semgrepImpact"])
}

func TestPreservesCrossFrameworkMetadata(t *testing.T) {
	out := convertFixture(t, "real.json")
	req := findReq(out.Baselines[0].Requirements, "dynamic-urllib-use-detected")
	require.NotNil(t, req)
	assert.Equal(t, "LOW", req.Tags["likelihood"])
	assert.Equal(t, "LOW", req.Tags["confidence"])
	assert.NotNil(t, req.Tags["asvs"])
	assert.NotNil(t, req.Tags["vulnerabilityClass"])
}

func TestReportsEveryFindingAsFailed(t *testing.T) {
	out := convertFixture(t, "real.json")
	for _, req := range out.Baselines[0].Requirements {
		for _, res := range req.Results {
			assert.Equal(t, hdf.Failed, res.Status)
		}
	}
}

func TestNeverEmitsRedactedPlaceholder(t *testing.T) {
	out := convertFixture(t, "real.json")
	for _, req := range out.Baselines[0].Requirements {
		for _, res := range req.Results {
			if res.Message != nil {
				assert.NotContains(t, *res.Message, redactedPlaceholder)
			}
		}
	}
}

func TestRecordsLocationInCodeDesc(t *testing.T) {
	out := convertFixture(t, "real.json")
	req := findReq(out.Baselines[0].Requirements, "subprocess-shell-true")
	require.NotNil(t, req)
	assert.Contains(t, req.Results[0].CodeDesc, "app/handlers.py")
	assert.Contains(t, req.Results[0].CodeDesc, "7")
}

func TestEmptyScanProducesNoFindingsRequirement(t *testing.T) {
	out := convertFixture(t, "empty.json")
	reqs := out.Baselines[0].Requirements
	require.Len(t, reqs, 1)
	assert.Equal(t, 0.0, reqs[0].Impact)
}

func TestScanErrorsBecomeTheirOwnRequirement(t *testing.T) {
	out := convertFixture(t, "errors.json")
	req := findReq(out.Baselines[0].Requirements, scanErrorsID)
	require.NotNil(t, req)
	assert.Len(t, req.Results, 2)
	for _, res := range req.Results {
		assert.Equal(t, hdf.Error, res.Status)
	}
}

func TestScanErrorsRequirementAbsentWhenNoErrors(t *testing.T) {
	out := convertFixture(t, "real.json")
	assert.Nil(t, findReq(out.Baselines[0].Requirements, scanErrorsID))
}

func TestTitleDerivedFromRuleId(t *testing.T) {
	assert.Equal(t, "Subprocess Shell True", titleFor("python.lang.security.audit.subprocess-shell-true.subprocess-shell-true"))
	assert.Equal(t, "Rule", titleFor("rule"))
}

func TestStringOrSliceDecodesBothForms(t *testing.T) {
	var s StringOrSlice
	require.NoError(t, s.UnmarshalJSON([]byte(`"one"`)))
	assert.Equal(t, StringOrSlice{"one"}, s)
	require.NoError(t, s.UnmarshalJSON([]byte(`["a","b"]`)))
	assert.Equal(t, StringOrSlice{"a", "b"}, s)
	require.NoError(t, s.UnmarshalJSON([]byte(`{"unexpected":true}`)))
	assert.Nil(t, s)
}

// ---- Golden snapshots (TS<->Go parity) ----

func TestSnapshots(t *testing.T) {
	shared.RunSnapshotTests(t, "semgrep-to-hdf", func(input []byte) (interface{}, error) {
		return ConvertSemgrepToHDF(input, testVersion)
	}, "*")
}

// ---- Edge cases ----

func TestImpactForFallsBackToModerate(t *testing.T) {
	assert.Equal(t, defaultImpact, impactFor(Result{}))
	assert.Equal(t, defaultImpact, impactFor(Result{Extra: Extra{Severity: "NOVEL"}}))
	assert.Equal(t, defaultImpact, impactFor(Result{Extra: Extra{Severity: redactedPlaceholder}}))
	assert.Equal(t, 0.9, impactFor(Result{Extra: Extra{Severity: "CRITICAL"}}))
}

func TestCodeDescVariants(t *testing.T) {
	assert.Equal(t, "Path: unknown", codeDescFor(Result{}))
	assert.Equal(t, "Path: a.py", codeDescFor(Result{Path: "a.py"}))
	assert.Equal(t, "Path: a.py, line 4",
		codeDescFor(Result{Path: "a.py", Start: Position{Line: 4}, End: Position{Line: 4}}))
	assert.Equal(t, "Path: a.py, lines 4-9",
		codeDescFor(Result{Path: "a.py", Start: Position{Line: 4}, End: Position{Line: 9}}))
}

func TestMessageForSkipsRedactedFields(t *testing.T) {
	assert.Equal(t, "", messageFor(Result{Extra: Extra{Lines: redactedPlaceholder}}))
	assert.Contains(t, messageFor(Result{Extra: Extra{Lines: "x = 1"}}), "Matched code:")
	assert.Contains(t, messageFor(Result{Extra: Extra{Fix: "False"}}), "replace the matched code with")
}

func TestErrorsRequirementHandlesOddTypeShapes(t *testing.T) {
	req := buildErrorsRequirement([]ScanError{
		{Message: "a", Type: []any{"PartialParsing", []any{}}},
		{Message: "b", Type: "PlainString"},
		{Message: "c"},
	}, time.Now().UTC())
	require.Len(t, req.Results, 3)
	assert.Contains(t, *req.Results[0].Message, "PartialParsing")
	assert.Contains(t, *req.Results[1].Message, "PlainString")
	assert.Contains(t, *req.Results[2].Message, "Unknown")
	assert.Equal(t, "Path: unknown", req.Results[2].CodeDesc)
}

func TestSkipsFindingsWithoutCheckId(t *testing.T) {
	input := []byte(`{"results":[{"path":"a.py"}],"errors":[],"paths":{"scanned":["a.py"]}}`)
	out, err := ConvertSemgrepToHDF(input, testVersion)
	require.NoError(t, err)
	// The malformed finding is skipped, leaving a no-findings placeholder.
	require.Len(t, out.Baselines[0].Requirements, 1)
	assert.Equal(t, "semgrep-no-findings", out.Baselines[0].Requirements[0].ID)
}

func TestReferencesDeduplicated(t *testing.T) {
	urls := referencesFor(Metadata{
		References: StringOrSlice{"https://a", "https://a"},
		Source:     "https://a",
		Shortlink:  "https://b",
		ASVS:       map[string]any{"control_url": "https://c"},
	})
	assert.Equal(t, []string{"https://a", "https://b", "https://c"}, urls)
}

func TestRejectsOversizedInput(t *testing.T) {
	_, err := ConvertSemgrepToHDF([]byte(`{"results":`), testVersion)
	require.Error(t, err)
}
